### PR TITLE
gRPC PHP Client Interceptor

### PR DESCRIPTION
**Pull request by ZhouyihaiDing**
This PR adds machinery to intercept RPCs on client side to gRPC PHP.

### CODE REVIEW

**Comment by ZhouyihaiDing** [comment]
Really appreciate @mehrdada for spending a lot of time helping to review the code before create this PR.
Going to write an gRFC.

**Comment by mehrdada** [comment]
Can you please change the commit message ("new implementation") to be more descriptive? (This is a good guide to follow: https://chris.beams.io/posts/git-commit/#seven-rules)

(For history preservation purposes, draft versions of this PR were reviewed https://github.com/ZhouyihaiDing/grpc/pull/4)

File: src/php/lib/Grpc/Interceptor.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                $channel = new InterceptorChannel($channel, $interceptors[$i]);
+            }
+        } else {
+            $channel =  new InterceptorChannel($channel, $interceptors);
+        }
+        return $channel;
+    }
+}
+
+class InterceptorChannel
```
Comment (mehrdada) [inline src/php/lib/Grpc/Interceptor.php]: If there is no way to make this private, please move it to a subdirectory named `Internal` and add a comment `This is a PRIVATE API and can change without notice.`

File: src/php/tests/unit_tests/InterceptorTest.php [hunk 2]
```diff
@@ -0,0 +1,401 @@
+<?php
+/*
+ *
+ * Copyright 2017 gRPC authors.
```
Comment (mehrdada) [inline src/php/tests/unit_tests/InterceptorTest.php]: Please update the date of the newly created files in this PR to 2018.

File: src/php/lib/Grpc/Interceptor.php (hunk 1 above)
Comment (ZhouyihaiDing) [reply src/php/lib/Grpc/Interceptor.php]: Done.

File: src/php/tests/unit_tests/InterceptorTest.php (hunk 2 above)
Comment (ZhouyihaiDing) [reply src/php/tests/unit_tests/InterceptorTest.php]: Done.

**Comment by mehrdada** [comment]
Ping

**Comment by hsaliak** [comment]
@ZhouyihaiDing @stanley-cheung what remains to be done for this PR?

It seems strange that the C/C++ build failed for the PHP only change. Should this just be a matter of re-running?

**Comment by ZhouyihaiDing** [comment]
@hsaliak I'll update and rebase it, and get it merged today.

**Comment by ZhouyihaiDing** [comment]
@mehrdada . PR is reviewed and rebased to one commit. Sorry about this delay!

**Comment by ZhouyihaiDing** [comment]
Interop Cloud-to-Cloud Tests (internal CI) - #15289
Interop Cloud-to-Prod Tests (internal CI) - #15289
Distribution Tests Linux (standalone subset)  - #14094

**Pull request merged**